### ISSUE TITLE
Warn on dyn AsyncTrait + Unpin

### ISSUE THREAD

**Issue by Darksonn** [rust-lang/rust-clippy#7603]
### What it does

This lint will detect usage of `dyn AsyncTrait + Unpin` for various different choices for `AsyncTrait`. It should also trigger when a `+ Send` is added.

The suggested replacement should be `Pin<Box<dyn AsyncTrait>>`. This is strictly more powerful than the `+ Unpin` version.

The lint should at least apply to these traits: `Future`, `Stream`, `AsyncRead`, `AsyncWrite`. It might even make sense to apply it to all traits.

### Categories (optional)

- Kind: style

The advantage is that changing the type like this allows storing non-`Unpin` types in the trait object.

### Drawbacks

None.

### Example

```rust
use futures::stream::Stream;

fn new_stream() -> Box<dyn Stream<Item = i32> + Unpin> {
    Box::new(futures::stream::iter(vec![1, 2, 3]))
}
```

Could be written as:

```rust
use futures::stream::Stream;

fn new_stream() -> Pin<Box<dyn Stream<Item = i32>>> {
    Box::pin(futures::stream::iter(vec![1, 2, 3]))
}
```
